### ISSUE TITLE
请问训练或者推理时如何自定义修改每个时间步的causal mask？

### ISSUE THREAD

**Issue by Aran-Guo** [QwenLM/Qwen1.5#455]
causal mask不是将pad token掩码掉的mask，而是每个时间步的可见区域，请问怎么修改实现？

**Comment by jklj077** [comment]
Hi, causal mask is basically implemented in all causal language models, Qwen1.5/2 included. Please refer to the code [here](https://github.com/huggingface/transformers/blob/96eb06286b63c9c93334d507e632c175d6ba8b28/src/transformers/models/qwen2/modeling_qwen2.py#L990-L1020).

In addition, the model can take in 3 types of attention mask: 
- `None` and it defaults to the causal mask
- bool tensor of dimension (batch_size, seq_length) and it will be converted to causal mask with proper padding mask
- bool tensor of dimension (batch_size, 1, seq_length_query, seq_length_key) and it will be used as it is, a non-causal mask, which typically should not be used for a causal language model unless you understand what you are doing

**Comment by Aran-Guo** [comment]
> Hi, causal mask is basically implemented in all causal language models, Qwen1.5/2 included. Please refer to the code [here](https://github.com/huggingface/transformers/blob/96eb06286b63c9c93334d507e632c175d6ba8b28/src/transformers/models/qwen2/modeling_qwen2.py#L990-L1020).
> 
> In addition, the model can take in 3 types of attention mask:
> 
> * `None` and it defaults to the causal mask
> * bool tensor of dimension (batch_size, seq_length) and it will be converted to causal mask with proper padding mask
> * bool tensor of dimension (batch_size, 1, seq_length_query, seq_length_key) and it will be used as it is, a non-causal mask, which typically should not be used for a causal language model unless you understand what you are doing

但是我应该怎么实现修改mask代码呢？这是我自定义AutoModelForCausalLM，然后from_pretrained千文模型，但是forward函数并没有被调用，这是怎么回事？
`class AutoModelForCausalLM1(AutoModelForCausalLM):
    def __init__(self, config):
        super().__init__(config)

    def forward(self, input_ids=None, **kwargs):
        print("forward")
        # 生成自定义掩码矩阵
        attention_mask = self._generate_custom_mask(input_ids, boundaries, mask_prob=1.0)

        # 调用原始模型的前向传播方法，传入修改后的掩码矩阵
        outputs = super().forward(input_ids=input_ids, attention_mask=attention_mask, **kwargs)
        return outputs`

**Comment by Aran-Guo** [comment]
> Hi, are you familar with `transformers` or Python? If not, I would suggest learning about them, as they not specific to Qwen.
> 
> There are many ways to implement a custom forward in transformers:
> 
> 1. Monkey patching: Implement a new forward function in your own code and `Qwen2ModelForCausalLM.forward = my_awesome_forward` before instanciating the class.
> 2. Source install: Clone the huggingface transformers repo, change the related code in modeling_qwen2.py and install the package, preferably with editable install.
> 3. Implement a new class: Implement a new class in your own code and then load the model with `MyAwesomeQwen2ModelForCausalLM.from_pretrained(...)`. (You cannot directly use the class with AutoModelForCausalLM. To register the class to be used with AutoModelForCausalLM, there are more to change, and please refer to the huggingface documentation for guidance.)
> 4. ...

好的，非常感谢，我试一下～
